pkg/services/object/transport_splitter.go

```diff
@@ -0,0 +1,210 @@
+)
+
+type (
+	TransportSplitter struct {
+		next object.Service
+
+		chunkSize  uint64
+		addrAmount uint64
+	}
+
+	getStreamBasicChecker struct {
+		next      object.GetObjectStreamer
+		buf       *bytes.Buffer
+		resp      *object.GetResponse
+		chunkSize uint64
+	}
+
+	searchStreamBasicChecker struct {
+		next       object.SearchObjectStreamer
+		resp       *object.SearchResponse
+		list       []*refs.ObjectID
+		addrAmount uint64
+	}
+
+	rangeStreamBasicChecker struct {
+		next      object.GetRangeObjectStreamer
+		buf       *bytes.Buffer
+		resp      *object.GetRangeResponse
+		chunkSize uint64
+	}
+)
+
+func NewTransportSplitter(size, amount uint64, next object.Service) *TransportSplitter {
+	return &TransportSplitter{
+		next:       next,
+		chunkSize:  size,
+		addrAmount: amount,
+	}
+}
+
+func (c TransportSplitter) Get(ctx context.Context, request *object.GetRequest) (object.GetObjectStreamer, error) {
+	stream, err := c.next.Get(ctx, request)
+
+	return &getStreamBasicChecker{
+		next:      stream,
+		chunkSize: c.chunkSize,
+	}, err
+}
+
+func (c TransportSplitter) Put(ctx context.Context) (object.PutObjectStreamer, error) {
+	return c.next.Put(ctx)
+}
+
+func (c TransportSplitter) Head(ctx context.Context, request *object.HeadRequest) (*object.HeadResponse, error) {
+	return c.next.Head(ctx, request)
+}
+
+func (c TransportSplitter) Search(ctx context.Context, request *object.SearchRequest) (object.SearchObjectStreamer, error) {
+	stream, err := c.next.Search(ctx, request)
+
+	return &searchStreamBasicChecker{
+		next:       stream,
+		addrAmount: c.addrAmount,
+	}, err
+}
+
+func (c TransportSplitter) Delete(ctx context.Context, request *object.DeleteRequest) (*object.DeleteResponse, error) {
+	return c.next.Delete(ctx, request)
+}
+
+func (c TransportSplitter) GetRange(ctx context.Context, request *object.GetRangeRequest) (object.GetRangeObjectStreamer, error) {
+	stream, err := c.next.GetRange(ctx, request)
+
+	return &rangeStreamBasicChecker{
+		next:      stream,
+		chunkSize: c.chunkSize,
+	}, err
+}
+
+func (c TransportSplitter) GetRangeHash(ctx context.Context, request *object.GetRangeHashRequest) (*object.GetRangeHashResponse, error) {
+	return c.next.GetRangeHash(ctx, request)
+}
+
+func (g *getStreamBasicChecker) Recv() (*object.GetResponse, error) {
+	if g.resp == nil {
+		resp, err := g.next.Recv()
+		if err != nil {
+			return resp, err
+		}
+
+		if part, ok := resp.GetBody().GetObjectPart().(*object.GetObjectPartChunk); !ok {
+			return resp, err
+		} else {
+			g.resp = resp
+			g.buf = bytes.NewBuffer(part.GetChunk())
+		}
+	}
+
+	chunk := make([]byte, g.chunkSize)
+	n, err := g.buf.Read(chunk)
+	if err != nil {
+		return nil, errors.Wrap(errChunking, err.Error())
+	}
+
+	chunkBody := new(object.GetObjectPartChunk)
+	chunkBody.SetChunk(chunk[:n])
+
+	body := new(object.GetResponseBody)
+	body.SetObjectPart(chunkBody)
+
+	resp := new(object.GetResponse)
```

Comment:
Unfortunately we can't. 

Consider there is a response that has to be split into two chunks. At first chunk, function returns `g.resp` to the sign service, which adds verification header to the response. Since `g.resp` is a pointer, the second chunk will have signature of the first chunk in `g.resp` and then it will be signed twice. 

In theory we can set verification header to the nil, but then it makes this service heavily depended on logic of surrounding services. This way it is independent, service just makes dup of the original response.